yourbench/pipeline/multi_hop_question_generation.py

```diff
@@ -0,0 +1,510 @@
+It processes a dataset of documents—each containing a list of multi-hop chunks—and generates
+multi-hop questions requiring integrative reasoning across those chunks. It uses a large
+language model to produce question-answer pairs in JSON format.
+
+Usage:
+------
+This module is typically invoked as part of the overall YourBench pipeline. It expects:
+1. A source dataset (e.g., documents with 'multihop_chunks' field).
+2. Configuration for multi-hop question generation, such as sampling parameters and
+   additional instructions.
+3. The pipeline orchestrator (in `handler.py`) calls `run(config)` if
+   `multi_hop_question_generation` is enabled in the YAML configuration.
+
+The module then:
+1. Optionally samples multi-hop chunks from each document.
+2. Prompts a large language model to generate multi-hop question-answer pairs.
+3. Parses and saves the generated questions in a structured dataset.
+
+Error Handling and Logging:
+---------------------------
+- Comprehensive logging is performed using `loguru` at various levels to trace execution.
+- Exceptions are caught and logged as errors, with the module attempting to continue
+  where practical.
+- Critical issues produce warnings or errors and gracefully terminate the stage.
+
+Module-Level Dependencies:
+--------------------------
+- Relies on the shared pipeline utilities (e.g., `yourbench.utils.dataset_engine`,
+  `yourbench.utils.inference_engine`, `yourbench.utils.prompts`).
+- Preserves the existing signature and functionality for downstream consistency.
+"""
+
+import json
+import random
+import re
+from dataclasses import dataclass, field
+from typing import Any, Dict, List
+
+from datasets import Dataset
+from loguru import logger
+
+from yourbench.utils.dataset_engine import (
+    save_dataset,
+    smart_get_output_dataset_name,
+    smart_get_output_subset,
+    smart_get_source_dataset_name,
+    smart_get_source_subset,
+    smart_load_dataset,
+)
+from yourbench.utils.inference_engine import InferenceCall, run_inference
+from yourbench.utils.prompts import (
+    MULTI_HOP_QUESTION_GENERATION_SYSTEM_PROMPT,
+    MULTI_HOP_QUESTION_GENERATION_USER_PROMPT,
+)
+
+
+@dataclass
+class MultiHopQuestionRow:
+    """
+    Data structure to represent a single multi-hop question row.
+
+    Attributes:
+        document_id (str):
+            Identifier of the document from which the question is generated.
+        source_chunk_ids (List[str]):
+            List of single-hop chunk IDs used in producing the multi-hop question.
+        question (str):
+            The generated multi-hop question text.
+        self_answer (str):
+            A plausible model-provided answer to the generated question.
+        estimated_difficulty (int):
+            Difficulty rating on a scale of 1 (easiest) to 10 (hardest).
+        self_assessed_question_type (str):
+            A descriptor for the question style (e.g., "analytical", "factual").
```

Comment:
provide full list of descriptors somewhere

Reply:
provided in the prompt, will also add in documentation